packages/SwingSet/docs/virtual-objects.md

```diff
@@ -0,0 +1,111 @@
+# Vat Secondary Storage
+
+The kernel currently provides two secondary storage mechanisms for the use of (application) code running in a vat:
+
+- Virtual objects
+- Persistent weak stores
+
+Each of these is accessed via a global made available to vat code.
+
+## Virtual Objects
+
+A virtual object is an object with a durable identity whose state is state is automatically and transparently backed up in secondary storage.  This means that when a given virtual object is not in active use, it need not occupy any memory within the executing vat.  Thus a vat can manage an arbitrarilly large number of such objects without concern about running out of memory.
+
+A virtual object has a "kind", which defines what sort of behavior and state it will possess.  You can think of a kind somewhat like a data type, but without direct linguistic support.
+
+A vat can define new kinds of virtual object by calling the `makeKind` function provided as a vat global:
+
+`makeKind(bodyMaker)`
+
+The return value from `makeKind` is a maker function which the vat can use to create instances of the newly defined object kind.
+
+The `bodyMaker` parameter is a user provided function that binds an in-memory instance of the virtual object with the virtual object's state, associating such instances with the virtual object's behavior.  It is called whenever a new virtual object instance is created, whenever such an instance is swapped in from secondary storage, and whenever a reference to a virtual object is received as a parameter of a message and deserialized.
+
+The `bodyMaker` function is passed the virtual object's state as a parameter, and is expected to return a new Javascript object with methods that close over this state.
+
+One distinguished method, named `initialize`, is called when new instances are first created, and is expected to initialize the state object.  Any parameters passed to the maker function returned by `makeKind` are passed directly to the `initialize` method.
```

Comment:
If we had some way to declare `initialize` as `private` that would do 90% of the job, as the main purpose of the magic is to ensure that handing out a reference to the object is not also handing out a capability to reset all its state.  If nobody tries to call `initialize` a second time they'll never notice the magic is even there.  (The other magic is the convention that the initialize method is named `initialize`, but that's more of an aesthetic concern.  I'd be open to marking it in some other way, though the specific suggestion of `__initialize__` is so horribly butt ugly it makes me want to scream.)

Also (and this is really more of a pedantic point) `initialize` _is_ a method of the actual object being made; it's just that it's removed after it's used for safety reasons.  In particular, if you had some kind of weird initialization dance that worked by having `initialize` hand the object to some external worker code that called `initialize` recursively, you could do that and it would work.

Reply:
> it's just that it's removed after it's used

With `delete`? God that's ugly. We should definitely find a way to express the initialization function other than by making it a method of the object.

> In particular, if you had some kind of weird initialization dance that worked by having initialize hand the object to some external worker code that called initialize recursively, you could do that and it would work.

That implies that you start executing before deleting it. That in turn implies that you start executing before hardening. We cannot do that.

Reply:
> > it's just that it's removed after it's used
>
> With delete? God that's ugly. We should definitely find a way to express the initialization function other than by making it a method of the object.

I'm open to any suggestion that preserves reasonable ergonomics.  Keep in mind that, as @warner has pointed out, the initializer needs to able to see the object that the body maker returns (or will return), not just the state, because it may need to pass it to somebody else as part of initialization.

> > In particular, if you had some kind of weird initialization dance that worked by having initialize hand the object to some external worker code that called initialize recursively, you could do that and it would work.
> 
> That implies that you start executing before deleting it. That in turn implies that you start executing before hardening. We cannot do that.

What do you mean by "start executing"?  Keep in mind that I'm talking here about what happens _during_ initialization, when the state is still malleable because it hasn't been entirely established yet.  We're still on the `initialize` call stack.

Reply:
> What do you mean by "start executing"?

That you call `initialize` while the object still has an `initialize` method. This is fatally dangerous if

>  initializer needs to able to see the object that the body maker returns (or will return), not just the state, because it may need to pass it to somebody else as part of initialization.

because then it is giving other things access to the object a) before it is hardened, and b) while it still has an `initialize` method that may be abused.

Reply:
I just realized a very important implementation detail that you missed (maybe because you're looking at the documentation PR instead of the code?) and that I straight up forgot (meaning, among other things, that my remarks above about whacky recursive initialization are just totally, utterly, and embarrassingly wrong): the `initialize` function is extracted and deleted from the object, and then the object is hardened, before `initialize` is called.   Nobody gets to see the object before it's hardened.

Reply:
Also, looking at the code, just found a bug in same, which I will now proceed to fix.  Grrr.